### ISSUE TITLE
`stress.py` still failing. No server-side timeouts, errors

### ISSUE THREAD

**Issue by jfalkner** [PacificBiosciences/smrtflow#205]
@skinner @mpkocher 

With today's code I'm no longer seeing `stress.py` pass. It sporadically fails. There appear to be no errors no timeouts in the logs. Instead I think it may be some sort of race condition where the status check occurs before the DB records the import.

Here is what I'm seeing in the logs for the failed import. After data import a GET is issued to ` http://localhost:8070/secondary-analysis/datasets/<dataset>`. When manually checked after the test, this URL shows the right result.

During the test, the one UUID that failed had this in the logs.

```
##$ cat server_out.txt | grep "1a569504-1c3c-e295-3af3-9a794140de12"
...
# HTTP GET is logged (response is not -- see next)
[info] 2016-06-24 18:28:54.421UTC INFO [smrtlink-analysis-server-akka.actor.default-dispatcher-16] a.a.RepointableActorRef - request: HttpRequest(GET,http://localhost:8070/secondary-analysis/datasets/1a569504-1c3c-e295-3af3-9a794140de12,List(Host: localhost:8070, Connection: keep-alive, Accept-Encoding: gzip, deflate, Accept: */*, User-Agent: python-requests/2.10.0, Content-Type: application/json),Empty,HTTP/1.1)

# 404 returned because import isn't done
[info] 2016-06-24  WARN [ForkJoinPool-2-worker-11] c.p.c.a.BaseApi$ServiceActor - Non-fatal exception: 404 - Not Found - Unable to find dataset 1a569504-1c3c-e295-3af3-9a794140de12 Request HttpRequest(GET,http://localhost:8070/secondary-analysis/datasets/1a569504-1c3c-e295-3af3-9a794140de12,List(Host: localhost:8070, Connection: keep-alive, Accept-Encoding: gzip, deflate, User-Agent: python-requests/2.10.0, Content-Type: application/json),Empty,HTTP/1.1)

# import ends after last attempted GET
[info] 2016-06-24  INFO [ForkJoinPool-2-worker-15] c.p.s.s.a.JobsDao - imported ReferencecSet 1a569504-1c3c-e295-3af3-9a794140de12 from /Users/jfalkner/tokeep/git/PacificBiosciences/smrtflow/test-output/dataset-159f0cd7-636d-4cb8-9d02-b8ce183ce251/referenceset.xml
```

Whereas all others that passed have a GET occur after the import succeeds.

```
# example of a successful import
$ cat server_out.txt | grep "a69b7789-3ca4-c2e7-21d2-28d37f069262"
...
# successful import
[info] 2016-06-24  INFO [ForkJoinPool-2-worker-15] c.p.s.s.a.JobsDao - imported ReferencecSet a69b7789-3ca4-c2e7-21d2-28d37f069262 from /Users/jfalkner/tokeep/git/PacificBiosciences/smrtflow/test-output/dataset-9f8f6908-6ecc-43bc-a15b-d657cd21b048/referenceset.xml

# GET that occurs after the import
[info] 2016-06-24 18:28:56.442UTC INFO [smrtlink-analysis-server-akka.actor.default-dispatcher-15] a.a.RepointableActorRef - request: HttpRequest(GET,http://localhost:8070/secondary-analysis/datasets/a69b7789-3ca4-c2e7-21d2-28d37f069262,List(Host: localhost:8070, Connection: keep-alive, Accept-Encoding: gzip, deflate, Accept: */*, User-Agent: python-requests/2.10.0, Content-Type: application/json),Empty,HTTP/1.1)
```

For a full run of stress.py, this was true for all of the below cases. 
[server_out.txt](https://github.com/PacificBiosciences/smrtflow/files/332406/server_out.txt)

Import Finished Before Final HTTP Request: http://localhost:8070/secondary-analysis/datasets/<dataset>
- 8b4db227-5a21-4658-8c1a-cf0f190d38e9
- f709059c-77a8-55c8-2408-17c6b07b04be
- f3c62360-e590-124d-6d4f-4d95cefb3242
- ed555362-6d46-bb68-15ee-e665c8c33131
- a69b7789-3ca4-c2e7-21d2-28d37f069262

Finished After
- 1a569504-1c3c-e295-3af3-9a794140de12

**Comment by jfalkner** [comment]
Yes, this appears to be a timing issue. 

https://github.com/PacificBiosciences/pbcommand/blob/master/pbcommand/services/service_access_layer.py#L492

Shows that there are two HTTP checks done. One before the import and one after. For some reason the import returns and the second check misses it.  A fix could be something like a server-side block to not return until the import is fully done or  adding a brief wait/retry to the python code.

**Comment by jfalkner** [comment]
> With today's code I'm no longer seeing stress.py pass. It sporadically fails.

Seems to pass  almost half the time.

**Comment by jfalkner** [comment]
@mpkocher and @natechols both noted that we should be using the scala version of the data import code. I'm unsure if this error would occur with that codebase.

**Comment by skinner** [comment]
In the testing on NFS that I've been doing, I'm pretty sure I'm using the python version of `pbservice import-dataset`; the output looks different between the original python `pbservice

**Comment by jfalkner** [comment]
Info on the root cause of this error -- assuming I correctly understand what `EngineJob` is doing. In short, it looks like the import URL endpoint returns once the DB has the respective `EngineJob` created. Not when the job completes. The RESTful check assumes that the job is done by the time it runs, which is a bad assumption.

GET hits [`createJob()` on  `ImportDataSetServiceType.scala:72`](https://github.com/PacificBiosciences/smrtflow/blob/3516762a7f1361ecc17b08c9cf8ae36dc0050f19/smrt-server-link/src/main/scala/com/pacbio/secondary/smrtlink/services/jobtypes/ImportDataSetServiceType.scala#L68-L79)

```
        post {
          optionalAuthenticate(authenticator.jwtAuth) { authInfo =>
            entity(as[ImportDataSetOptions]) { sopts =>
              complete {
                created {
                  createJob(sopts, authInfo.map(_.login)) # HTTP entry point
```

That ends up eventually running [`createJob()` in `JobsDao.scala:327`](https://github.com/PacificBiosciences/smrtflow/blob/3516762a7f1361ecc17b08c9cf8ae36dc0050f19/smrt-server-link/src/main/scala/com/pacbio/secondary/smrtlink/actors/JobsDao.scala#L327-L367), which returns once this [`DBIO.seq` ends after creating the `EngineJob` in the DB](https://github.com/PacificBiosciences/smrtflow/blob/3516762a7f1361ecc17b08c9cf8ae36dc0050f19/smrt-server-link/src/main/scala/com/pacbio/secondary/smrtlink/actors/JobsDao.scala#L359-L366).

```
  createJob(...)
    ...
     DBIO.seq(
        engineJobs.filter(_.id === jobId).map(_.path).update(resolvedPath),
        jobEvents += jobEvent,
        engineJobsDataSets ++= entryPoints.getOrElse(Nil).map(e => EngineJobEntryPoint(jobId, e.datasetUUID, e.datasetType))
      ).map(_ => engineJob.copy(id = jobId, path = resolvedPath))
    }

    db.run(updates.transactionally)
   # at this point the engine job is crated but may not be done
  }

**Comment by jfalkner** [comment]
From a chat with @mpkocher, above is the expected behavior. The glitch must then be on the python side. I'll trace that further.

**Comment by jfalkner** [comment]
The code looks like it'll retry every 2 seconds up to 5 minutes. There are also two validation checks. It'd be helpful to confirm which ones are observed.

https://github.com/PacificBiosciences/pbcommand/blob/master/pbcommand/services/service_access_layer.py#L480-L483

or 

https://github.com/PacificBiosciences/pbcommand/blob/master/pbcommand/services/service_access_layer.py#L494-L497

@skinner I had deleted my `stress_err.txt` and `stress_out.txt` logs and can't seem to replicate it against locally. I just ran the `stress.py` test a half a dozen times with no issues. I think our codebase knows it is the weekend and 5pm in my timezone.

More seriously, earlier in the day I had Pandora playing music and was multi-tasking. Leads me to suspect I'll need to bog my machine down some to replicate this again. That also supports that it is a timing issue somewhere.
